### ISSUE TITLE
[Bug] Flashinfer backend get wrong result on last layer of qwen2 model

### ISSUE THREAD

**Issue by jianfei-wangg** [sgl-project/sglang#13044]
### Checklist

- [x] 1. I have searched related issues but cannot get the expected help.
- [x] 2. The bug has not been fixed in the latest version.
- [x] 3. Please note that if the bug-related issue you submitted lacks corresponding environment info and a minimal reproducible demo, it will be challenging for us to reproduce and resolve the issue, reducing the likelihood of receiving feedback.
- [x] 4. If the issue you raised is not a bug but a question, please raise a discussion at https://github.com/sgl-project/sglang/discussions/new/choose Otherwise, it will be closed.
- [x] 5. Please use English, otherwise it will be closed.

### Describe the bug

I have encountered qwen2model accuracy problem, and I found it can be reproduced by compare scaled_dot_product_attention with sglang's attention_backend, fa3 or triton has no problem.

### Reproduction

here is the code snippet:
`class Qwen2Attention(nn.Module):
    def __init__(
        self,
        hidden_size: int,
        num_heads: int,
        num_kv_heads: int,
        head_dim: Optional[int] = None,
        layer_id: int = 0,
        rope_theta: float = 1000000,
        rope_scaling: Optional[Dict[str, Any]] = None,
        max_position_embeddings: int = 32768,
        quant_config: Optional[QuantizationConfig] = None,
        dual_chunk_attention_config: Optional[dict[str, Any]] = None,
        prefix: str = "",
    ) -> None:
        super().__init__()
        self.hidden_size = hidden_size
        tp_size = get_tensor_model_parallel_world_size()
        self.total_num_heads = num_heads
        assert self.total_num_heads % tp_size == 0
        self.num_heads = self.total_num_heads // tp_size
        self.total_num_kv_heads = num_kv_heads
        if self.total_num_kv_heads >= tp_size:
            # Number of KV heads is greater than TP size, so we partition
            # the KV heads across multiple tensor parallel GPUs.
            assert self.total_num_kv_heads % tp_size == 0
        else:
            # Number of KV heads is less than TP size, so we replicate
            # the KV heads across multiple tensor parallel GPUs.
            assert tp_size % self.total_num_kv_heads == 0
        self.num_kv_heads = max(1, self.total_num_kv_heads // tp_size)
        if head_dim is not None:
            self.head_dim = head_dim
        else:
            self.head_dim = hidden_size // self.total_num_heads
        self.q_size = self.num_heads * self.head_dim
        self.kv_size = self.num_kv_heads * self.head_dim
        self.scaling = self.head_dim**-0.5
        self.rope_theta = rope_theta
        self.max_position_embeddings = max_position_embeddings
        self.layer_id = layer_id

        self.qkv_proj = QKVParallelLinear(
            hidden_size,
            self.head_dim,
            self.total_num_heads,
            self.total_num_kv_heads,
            bias=True,
            quant_config=quant_config,
            prefix=add_prefix("qkv_proj", prefix),
        )
        self.o_proj = RowParallelLinear(
            self.total_num_heads * self.head_dim,
            hidden_size,
            bias=False,
            quant_config=quant_config,
            prefix=add_prefix("o_proj", prefix),
        )

        self.rotary_emb = get_rope(
            self.head_dim,
            rotary_dim=self.head_dim,
            max_position=max_position_embeddings,
            base=rope_theta,
            rope_scaling=rope_scaling,
            dual_chunk_attention_config=dual_chunk_attention_config,
        )
        self.attn = RadixAttention(
            self.num_heads,
            self.head_dim,
            self.scaling,
            num_kv_heads=self.num_kv_heads,
            layer_id=layer_id,
            quant_config=quant_config,
            prefix=add_prefix("attn", prefix),
        )

    def _scaled_dot_product_attention(self, q, k, v):
        import torch
        seq_len = q.shape[0]
        q_reshaped = q.view(seq_len, self.num_heads, self.head_dim).unsqueeze(0).transpose(1, 2)
        k_reshaped = k.view(seq_len, self.num_kv_heads, self.head_dim).unsqueeze(0).transpose(1, 2)
        v_reshaped = v.view(seq_len, self.num_kv_heads, self.head_dim).unsqueeze(0).transpose(1, 2)
        if self.num_kv_heads != self.num_heads:
            k_reshaped = k_reshaped.repeat_interleave(self.num_heads // self.num_kv_heads, dim=1)
            v_reshaped = v_reshaped.repeat_interleave(self.num_heads // self.num_kv_heads, dim=1)
        attn_output_scaled = torch.nn.functional.scaled_dot_product_attention(q_reshaped, k_reshaped, v_reshaped, is_causal=True)
        return attn_output_scaled.transpose(1, 2).squeeze(0).reshape(seq_len, -1)
    
    def _compare_two_tensors(self, tensor1, tensor2):
        import torch
        diff = tensor1 - tensor2
        max_diff = torch.max(torch.abs(diff))
        mean_diff = torch.mean(torch.abs(diff))
        cosine_sim = torch.nn.functional.cosine_similarity(tensor1, tensor2, dim=-1)
        print(f"layer_{self.layer_id} Max diff: {max_diff}, Mean diff: {mean_diff}, Cosine sim: {cosine_sim}")

    def forward(
        self,
        positions: torch.Tensor,
        hidden_states: torch.Tensor,
        forward_batch: ForwardBatch,
    ) -> torch.Tensor:
        qkv, _ = self.qkv_proj(hidden_states)
        q, k, v = qkv.split([self.q_size, self.kv_size, self.kv_size], dim=-1)
        q, k = self.rotary_emb(positions, q, k)
        attn_output = self.attn(q, k, v, forward_batch)
        if hidden_states.shape[0] > 1:
            scaled_attn_output = self._scaled_dot_product_attention(q, k, v)
            self._compare_two_tensors(attn_output, scaled_attn_output)
        output, _ = self.o_proj(attn_output)
        return output`
and then run:
`python -m sglang.launch_server --model-path Qwen/Qwen2-7B-Instruct --context-length 32768 --max-running-requests 32 --tp-size 1 --port 8192 --cuda-graph-max-bs 32 --mem-fraction-static 0.85 --trust-remote-code --chunked-prefill-size -1 --log-level info --disable-cuda-graph --attention-backend flashinfer --host '0.0.0.0'`
you will get:

<img width="1972" height="450" alt="Image" src="https://github.com/user-attachments/assets/b56d902a-2536-4c23-9ab4-644684f23667" />

### Environment

Python: 3.12.11 (main, Jun 12 2025, 02:44:25) [GCC 13.3.0]
CUDA available: True
GPU 0,1,2,3,4,5,6,7: NVIDIA L20Z
GPU 0,1,2,3,4,5,6,7 Compute Capability: 9.0
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 12.6, V12.6.85
CUDA Driver Version: 550.127.08
PyTorch: 2.8.0+cu128
sglang: 0.5.5.post1
sgl_kernel: 0.3.17
flashinfer_python: 0.5.0
flashinfer_cubin: 0.5.0
flashinfer_jit_cache: Module Not Found
triton: 3.4.0
transformers: 4.57.1
torchao: 0.9.0
numpy: 2.3.3
aiohttp: 3.12.15
fastapi: 0.116.1
hf_transfer: 0.1.9
huggingface_hub: 0.34.4
interegular: 0.3.3
modelscope: 1.29.2
orjson: 3.11.3
outlines: 0.1.11
packaging: 25.0
psutil: 7.0.0
pydantic: 2.11.7
python-multipart: 0.0.20
pyzmq: 27.1.0
uvicorn: 0.35.0
uvloop: 0.21.0
vllm: Module Not Found
xgrammar: 0.1.25
openai: 2.6.1
tiktoken: 0.11.0
anthropic: 0.67.0
litellm: Module Not Found
decord2: 2.0.0
NVIDIA Topology: 
        GPU0    GPU1    GPU2    GPU3    GPU4    GPU5    GPU6    GPU7    NIC0    NIC1    NIC2    NIC3    NIC4    NIC5    NIC6    NIC7    CPU Affinity    NUMA Affinity   GPU NUMA ID
GPU0     X      NV18    NV18    NV18    NV18    NV18    NV18    NV18    PIX     PHB     PHB     PHB     SYS     SYS     SYS     SYS     0-63    0               N/A
GPU1    NV18     X      NV18    NV18    NV18    NV18    NV18    NV18    PHB     PIX     PHB     PHB     SYS     SYS     SYS     SYS     0-63    0               N/A
GPU2    NV18    NV18     X      NV18    NV18    NV18    NV18    NV18    PHB     PHB     PIX     PHB     SYS     SYS     SYS     SYS     0-63    0               N/A
GPU3    NV18    NV18    NV18     X      NV18    NV18    NV18    NV18    PHB     PHB     PHB     PIX     SYS     SYS     SYS     SYS     0-63    0               N/A
GPU4    NV18    NV18    NV18    NV18     X      NV18    NV18    NV18    SYS     SYS     SYS     SYS     PIX     PHB     PHB     PHB     64-127  1               N/A
GPU5    NV18    NV18    NV18    NV18    NV18     X      NV18    NV18    SYS     SYS     SYS     SYS     PHB     PIX     PHB     PHB     64-127  1               N/A
GPU6    NV18    NV18    NV18    NV18    NV18    NV18     X      NV18    SYS     SYS     SYS     SYS     PHB     PHB     PIX     PHB     64-127  1               N/A
GPU7    NV18    NV18    NV18    NV18    NV18    NV18    NV18     X      SYS     SYS     SYS     SYS     PHB     PHB     PHB     PIX     64-127  1               N/A
NIC0    PIX     PHB     PHB     PHB     SYS     SYS     SYS     SYS      X      PHB     PHB     PHB     SYS     SYS     SYS     SYS
NIC1    PHB     PIX     PHB     PHB     SYS     SYS     SYS     SYS     PHB      X      PHB     PHB     SYS     SYS     SYS     SYS
NIC2    PHB     PHB     PIX     PHB     SYS     SYS     SYS     SYS     PHB     PHB      X      PHB     SYS     SYS     SYS     SYS
NIC3    PHB     PHB     PHB     PIX     SYS     SYS     SYS     SYS     PHB     PHB     PHB      X      SYS     SYS     SYS     SYS
NIC4    SYS     SYS     SYS     SYS     PIX     PHB     PHB     PHB     SYS     SYS     SYS     SYS      X      PHB     PHB     PHB
NIC5    SYS     SYS     SYS     SYS     PHB     PIX     PHB     PHB     SYS     SYS     SYS     SYS     PHB      X      PHB     PHB
NIC6    SYS     SYS     SYS     SYS     PHB     PHB     PIX     PHB     SYS     SYS     SYS     SYS     PHB     PHB      X      PHB
NIC7    SYS     SYS     SYS     SYS     PHB     PHB     PHB     PIX     SYS     SYS     SYS     SYS     PHB     PHB     PHB      X 

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NIC Legend:

  NIC0: mlx5_0
  NIC1: mlx5_1
  NIC2: mlx5_2
  NIC3: mlx5_3
  NIC4: mlx5_4
  NIC5: mlx5_5
  NIC6: mlx5_6
  NIC7: mlx5_7

Hypervisor vendor:: KVM
ulimit soft: 102400

**Comment by i-love-debugging** [comment]
Hi,

We also did a diagnosis on this issue and could confirm the same behavior as listed in the original issue. We ran an example prompt from GSM8k benchmark on SGLang v0.5.6 and compared the activations of torch submodules in the Qwen model. Below listed the normalized error ratio calculated using L2-Norm on the output activations from HF and SGLang.


### SGLang FlashInfer vs HF Error Ratio

```
    [
        "root.model.layers.27.self_attn.qkv_proj",
        0.7378732141329455
    ],
    [
        "root.model.layers.27.self_attn.rotary_emb",
        false
    ],
    [
        "root.model.layers.27.self_attn.attn",
        false
    ],
    [
        "root.model.layers.27.self_attn.o_proj",
        10.18625227703632
    ],
    [
        "root.model.layers.27.self_attn",
        10.18625227703632
    ]

```


### SGLang FA3 vs HF Error Ratio

```
    [
        "root.model.layers.27.self_attn.qkv_proj",
        0.7378732141329455
    ],
    [
        "root.model.layers.27.self_attn.rotary_emb",
        false
    ],
    [
        "root.model.layers.27.self_attn.attn",
        false
    ],
    [
        "root.model.layers.27.self_attn.o_proj",
        1.0966950444254433
    ],
    [
        "root.model.layers.27.self_attn",
        1.0966950444254433
    ]
```

### SGLang Triton vs HF Error Ratio

```
    [
        "root.model.layers.27.self_attn.qkv_proj",
        0.7271793994353666
    ],
    [
        "root.model.layers.27.self_attn.rotary_emb",
        false
    ],
    [
        "root.model.layers.27.self_attn.attn",
        false
    ],
    [
        "root.model.layers.27.self_attn.o_proj",
        1.0966950444254433
    ],
    [
        "root.model.layers.27.self_attn",
        1.0966950444254433
    ],
```

We can see that after the last layer attention, the error ratio drastically increase for SGLang FlashInfer, compared to SGLang FA3 and SGLang Triton, indicating a bug.

We tried to find the root cause by following the code that initiates the attention backend. We did not find any obvious bugs in torch level code. So the issue seems to come from Flashinfer kernel that is JIT compiled.

Also interestingly, Qwen3 does not have such an issue for SGLang FlashInfer.

**Comment by gaoyang07** [comment]
@jianfei-wangg @i-love-debugging Any feedback on this issue?
